File: manejoProyectos/trunk/src/ar/edu/design/examples/manejoProyectos/Tarea.java
package ar.edu.design.examples.manejoProyectos;

import java.util.ArrayList;
import java.util.Collection;

public class Tarea {

	private Complejidad complejidad;
	private Collection<Impuesto> impuestos;
	private int tiempo;
	private TipoDeTarea tipoDeTarea;

	public Tarea() {
		this.complejidad = new ComplejidadMinima();
		this.impuestos = new ArrayList<Impuesto>();
		this.tiempo = 0;
		this.tipoDeTarea = new TareaSimple();
	}
	
	public double getCosto() {
		return tipoDeTarea.getCosto(this);
	}

	public double getCostoTotal() {
		return tipoDeTarea.getCostoTotal(this);
	}
	
	public double getCostoBase() {
		double costo = this.complejidad.getCosto(this);
		return costo + this.getCostoImpositivo(costo);	
	}

	public double getCostoImpositivo(double costo) {
		double total = 0;
		for (Impuesto impuesto : this.impuestos) {
			total += impuesto.getCostoImpositivo(costo);
		}
		return total;
	}

	public TipoDeTarea getTipoDeTarea() {
		return tipoDeTarea;
	}

	public void setCompuesta() {
		this.tipoDeTarea = new TareaCompuesta();
	}

	public void setSimple() {
		this.tipoDeTarea = new TareaSimple();
	}

	public int getTiempo() {
		return tiempo;
	}

	public void setTiempo(int tiempo) {
		this.tiempo = tiempo;
	}

	public void setComplejidad(Complejidad complejidad) {
		this.complejidad = complejidad;
	}

	public void agregarSubtarea(Tarea tarea) {
		this.tipoDeTarea.agregarSubtarea(tarea);
	}

}

File: funcional/manejo-proyectos-funcional/src/scala/proyectosv1/Tarea.scala
package proyectosv1

trait Tarea {
  def costo: BigDecimal
  def duracionEnDias: Int

  def costoBase = duracionEnDias * 25
}

class TareaSimple(
  val duracionEnDias: Int,
  complejidad: Complejidad,
  impuestos: Collection[Impuesto]) extends Tarea {

  def costo = complejidad.costoTarea(this) * (1 + porcentajeImpuestos / 100)
  def porcentajeImpuestos = impuestos.view.map(_.porcentajeImpuesto).sum
}

class TareaCompuesta(tareas: Collection[Tarea]) extends Tarea {
  def costo = tareas.view.map(_.costo).sum
  def duracionEnDias = tareas.view.map(_.duracionEnDias).sum
}